### ISSUE TITLE
Copy scene tree as text

### ISSUE THREAD

**Issue by theraot** [godotengine/godot-proposals#2953]
### Describe the project you are working on

None in particular.

### Describe the problem or limitation you are having in your project

Godot does not include a feature to copy the scene tree as text.

As a: Beginner.
I want to: Copy the scene tree as text.
So that: I can paste it on sites where screenshots are not supported or are frown upon. To ease asking questions.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

A new option in the contextual menu (secondary click) for nodes in the Scene panel to copy the branch (the node on which the the contextual menu was opened and its children, recursively) as text.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

- The user right clicks (or whatever the secondary click is mapped to) on a node on the scene tree
- The contextual menu opens up
- The user clicks on "Copy as text" (or other wording)
- Godot puts a text only representation of the tree on the clipboard.

The text representation might look something like this:

```
Enemy: KinematicBody2D (Enemey.gd)
├ Sprite
├ CollisionShape
└ StompDetector: Area2D
   └ CollisionShape2D
```

That is "Ascii art" of the tree - I'm suggesting using box-drawing characters - where each node represented by its name, the node type and the script script attached to it. Plus, I believe the node type can be omitted if it is evident from the name (i.e. when the name of the node includes the name of type of the node).

### If this enhancement will not be used often, can it be worked around with a few lines of script?

I believe the scene tree is accesible via add-on. I'm not sure how extensive the code would be.

### Is there a reason why this should be core and not an add-on in the asset library?

This would ease beginners asking questions in site that prefer text, do not support images, or frown upon images. Such as godotengine.org/qa, stackoverflow.com and gamedev.stackexchange.com - Thus, I believe it should be available without the need of an add-on.

To copy and paste the code of a relevant script is not issue at all. However, often scripts reference other nodes, and it can be difficult or impossible to answer question is the situation where the relevant portion of the scene tree is not included (which is something a beginner might not realize). Somewhat more experienced developers would take steps to produce such text representations of the scene tree (there are many examples in the aforementioned sites). I would like to see the friction to generate them reduced. I want to be able to describe the steps to produce such tree to a beginner in a few words (e.g. "Right click on the node, select 'copy as text' and paste it on the question").

**Comment by Calinou** [comment]
This can be done using `print_tree_pretty()`, although it will not display node types next to node names. (I guess that could be added.)

**Comment by MaaaxiKing** [comment]
I guess it's too much effort to write a script in which you call that functio. Therefore, a new item in the rightclick-dialog should exist: `Copy Tree`. Maybe also `Copy Tree from here`.

**Comment by theraot** [comment]
@MaaaxiKing There is already a "Copy" option, which is used to copy nodes around. Which is why I suggest "Copy as text".

I believe usually it is much less effort to type the tree than to use `print_tree_pretty` to get it. Regardless, I don't expect a beginner to know how to do that. And I'm not sure how I would explain it in the short space of a comment of Q&A sites.

Instead, I take @Calinou comment as a hint on where to begin to implement this feature.

**Comment by MaaaxiKing** [comment]
Yeah, then "Copy as text", I'd like to have it without a script anyways.

**Comment by theraot** [comment]
I've been thinking for these functions including the script or not, and including type or not, could be specified with a flag parameter.

Once the type is not a permanent fixture, I would expect omitting if the name includes it to be less trouble to implement… Yet, one you open the door for there being option, why that not an option? But then it isn't exactly flags, isn't it? I don't know. 

Regardless, I care more about it being in the context menu over whether or not it includes types or script. And I'm not sure exposing such options in a dialog is a good idea anyway.

**Comment by joulupukk1** [comment]
Hello, is there any update on this? To me it would be a huge QOL improvement. Thank you.

**Comment by arkology** [comment]
- See https://github.com/godotengine/godot/pull/119652
